src/django_components/context.py

```diff
@@ -5,31 +5,74 @@
 You can think of the Context as our storage system.
 """
 
-from copy import copy
 from typing import TYPE_CHECKING, Optional
 
 from django.template import Context
 
 from django_components.logger import trace_msg
+from django_components.utils import find_last_index
 
 if TYPE_CHECKING:
     from django_components.slots import FillContent
 
 
 _FILLED_SLOTS_CONTENT_CONTEXT_KEY = "_DJANGO_COMPONENTS_FILLED_SLOTS"
-_OUTER_CONTEXT_CONTEXT_KEY = "_DJANGO_COMPONENTS_OUTER_CONTEXT"
+_OUTER_ROOT_CTX_CONTEXT_KEY = "_DJANGO_COMPONENTS_OUTER_ROOT_CTX"
 _SLOT_COMPONENT_ASSOC_KEY = "_DJANGO_COMPONENTS_SLOT_COMP_ASSOC"
 
 
+def prepare_context(context: Context, outer_context: Optional[Context]) -> None:
+    """Initialize the internal context state."""
+    # This is supposed to run ALWAYS at Component.render
+    if outer_context is not None:
+        set_outer_root_context(context, outer_context)
+
+    # Initialize mapping dicts within this rendering run.
+    # This is shared across the whole render chain, thus we set it only once.
+    if _SLOT_COMPONENT_ASSOC_KEY not in context:
+        context[_SLOT_COMPONENT_ASSOC_KEY] = {}
+    if _FILLED_SLOTS_CONTENT_CONTEXT_KEY not in context:
+        context[_FILLED_SLOTS_CONTENT_CONTEXT_KEY] = {}
+
+    # If we're inside a forloop, we need to make a disposable copy of slot -> comp
+    # mapping, which can be modified in the loop. We do so by copying it onto the latest
+    # context layer.
+    #
+    # This is necessary, because otherwise if we have a nested loop with a same
+    # component used recursively, the inner slot -> comp mapping would leak into the outer.
+    #
+    # NOTE: If you ever need to debug this, insert a print/debug statement into
+    # `django.template.defaulttags.ForNode.render` to inspect the context object
+    # inside the for loop.
+    if "forloop" in context:
+        context.dicts[-1][_SLOT_COMPONENT_ASSOC_KEY] = context[_SLOT_COMPONENT_ASSOC_KEY].copy()
+
+
+def make_isolated_context_copy(context: Context) -> Context:
+    # Even if contexts are isolated, we still need to pass down the
+    # metadata so variables in slots can be rendered using the correct context.
+    root_ctx = get_outer_root_context(context)
+    slot_assoc = context.get(_SLOT_COMPONENT_ASSOC_KEY, {})
+    slot_fills = context.get(_FILLED_SLOTS_CONTENT_CONTEXT_KEY, {})
+
+    context_copy = context.new()
+    context_copy[_SLOT_COMPONENT_ASSOC_KEY] = slot_assoc
+    context_copy[_FILLED_SLOTS_CONTENT_CONTEXT_KEY] = slot_fills
+    set_outer_root_context(context_copy, root_ctx)
+    copy_forloop_context(context, context_copy)
```

Comment:
Also had to do pass down the "outer context" and for loop info for things to work